### PR TITLE
Add utility to use `HabanaDataLoader`

### PR DESCRIPTION
**Pull request by jerome-habana**
## What does this PR do?
This PR adds support to exercise habana daloader through the PTL interfaces. The current supported dataset using habana dataloader is only Imagenet.

[Reference](https://docs.habana.ai/en/latest/PyTorch/Habana_Media_Loader_PT/Media_Loader_PT.html)

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source-pytorch/accelerators/hpu_intermediate.rst [hunk 1]
```diff
@@ -97,3 +97,60 @@ The below snippet shows how DeviceStatsMonitor can be enabled.
     trainer = Trainer(accelerator="hpu", callbacks=[device_stats])
 
 For more details, please refer to `Memory Stats APIs <https://docs.habana.ai/en/v1.5.0/PyTorch/PyTorch_User_Guide/Python_Packages.html#memory-stats-apis>`__.
+
+----
+
+Using HPU Data Module
+----------------------
```
Comment (kaushikb11) [inline docs/source-pytorch/accelerators/hpu_intermediate.rst:104]:
```suggestion
Using HPU Data Module
------------------------
```

File: tests/tests_pytorch/accelerators/test_hpu.py [hunk 2]
```diff
@@ -323,3 +325,93 @@ def test_hpu_device_stats_monitor(tmpdir):
     ]
     for f in fields:
         assert any(f in h for h in hpu_stats.keys())
+
+
+@RunIf(hpu=True)
+def test_hpu_datamodule():
```
Comment (kaushikb11) [inline tests/tests_pytorch/accelerators/test_hpu.py:331]: let's also add a test where it works

Comment (kaushikb11) [reply tests/tests_pytorch/accelerators/test_hpu.py:331]: But it only seems to support ImageNet dataset, right?

File: docs/source-pytorch/accelerators/hpu_intermediate.rst (hunk 1 above)
Comment (kaushikb11) [inline docs/source-pytorch/accelerators/hpu_intermediate.rst:104]:
```suggestion
Using HPUDataModule
-----------------------
```

File: docs/source-pytorch/accelerators/hpu_intermediate.rst [hunk 3]
```diff
@@ -97,3 +97,60 @@ The below snippet shows how DeviceStatsMonitor can be enabled.
...
+
+HPUDataModule class is a wrapper around the pl.LightningDataModule class. It makes working with custom models easier on HPU devices.
+It uses HabanaDataloader for training, test and validation of user-provided models. Currently, it only supports Imagenet dataset.
```
Comment (kaushikb11) [inline docs/source-pytorch/accelerators/hpu_intermediate.rst:107]:
```suggestion
``HPUDataModule`` class is a wrapper around the ``LightningDataModule`` class. It makes working with custom models easier on HPU devices.
It uses HabanaDataloader for training, testing, and validation of user-provided models. Currently, it only supports the ``Imagenet`` dataset.
```

File: docs/source-pytorch/accelerators/hpu_intermediate.rst [hunk 4]
```diff
@@ -97,3 +97,60 @@ The below snippet shows how DeviceStatsMonitor can be enabled.
...
+
+The below snippet shows an example of how to use HPUDataModule.
```
Comment (kaushikb11) [inline docs/source-pytorch/accelerators/hpu_intermediate.rst:109]:
```suggestion
The below snippet shows an example of how to use the ``HPUDataModule``.
```

File: tests/tests_pytorch/accelerators/test_hpu.py (hunk 2 above)
Comment (jerome-habana) [reply tests/tests_pytorch/accelerators/test_hpu.py:333]: 1.5 release has imagenet support only. Also, there is an associated test in example

File: src/pytorch_lightning/CHANGELOG.md [hunk 5]
```diff
@@ -153,6 +153,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Added support for DDP Fork ([#13405](https://github.com/Lightning-AI/lightning/pull/13405))
 - Added support for async checkpointing ([#13658](https://github.com/Lightning-AI/lightning/pull/13658))
 - Added support for HPU Device stats monitor ([#13819](https://github.com/Lightning-AI/lightning/pull/13819))
+- Added support for HPU DataModule ([#13995](https://github.com/Lightning-AI/lightning/pull/13995))
```
Comment (kaushikb11) [inline src/pytorch_lightning/CHANGELOG.md:156]: The entry should be added for `1.8.0` section

File: docs/source-pytorch/accelerators/hpu_intermediate.rst [hunk 6]
```diff
@@ -97,3 +97,60 @@ The below snippet shows how DeviceStatsMonitor can be enabled.
...
+
+----
+
+Using HPUDataModule
+-----------------------
+
+``HPUDataModule`` class is a wrapper around the ``LightningDataModule`` class. It makes working with custom models easier on HPU devices.
+It uses HabanaDataloader for training, testing, and validation of user-provided models. Currently, it only supports the ``Imagenet`` dataset.
+
+The below snippet shows an example of how to use the ``HPUDataModule``.
```
Comment (Felonious-Spellfire) [inline docs/source-pytorch/accelerators/hpu_intermediate.rst:109]:
```suggestion
Here's an example of how to use the ``HPUDataModule``:
```

File: docs/source-pytorch/accelerators/hpu_intermediate.rst [hunk 7]
```diff
@@ -97,3 +97,60 @@ The below snippet shows how DeviceStatsMonitor can be enabled.
...
+    # Initialize a trainer
+    trainer = pl.Trainer(devices=1, accelerator="hpu", max_epochs=1, max_steps=2)
+
+    # Init our model
+    model = RN50Module()  # Or any other model to be defined by user
+
+    trainer.fit(model, datamodule=data_module)
+    trainer.validate(model, datamodule=data_module)
+
+A working example can be found at examples/pl_hpu/hpu_datamodule_sample.py
```
Comment (Felonious-Spellfire) [inline docs/source-pytorch/accelerators/hpu_intermediate.rst:155]:
```suggestion
A working example can be found at ``examples/pl_hpu/hpu_datamodule_sample.py``.
```

File: docs/source-pytorch/accelerators/hpu_intermediate.rst [hunk 8]
```diff
@@ -97,3 +97,60 @@ The below snippet shows how DeviceStatsMonitor can be enabled.
...
+For more details please refer to `Habana dataloader <https://docs.habana.ai/en/latest/PyTorch_User_Guide/PyTorch_User_Guide.html#habana-data-loader>`__.
```
Comment (Felonious-Spellfire) [inline docs/source-pytorch/accelerators/hpu_intermediate.rst:156]:
```suggestion
For more details refer to `Habana dataloader <https://docs.habana.ai/en/latest/PyTorch_User_Guide/PyTorch_User_Guide.html#habana-data-loader>`__.
```

File: src/pytorch_lightning/utilities/hpu_datamodule.py [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+    def setup(self, stage: Optional[str] = None):  # type: ignore[no-untyped-def]
+
+        if not _TORCHVISION_AVAILABLE:
+            raise ValueError("torchvision transforms not available")
+
+        if self.shuffle is True:
+            raise ValueError("HabanaDataLoader does not support shuffle=True")
+
+        if self.pin_memory is False:
+            raise ValueError("HabanaDataLoader supports only pin_memory=True")
```
Comment (Felonious-Spellfire) [inline src/pytorch_lightning/utilities/hpu_datamodule.py:123]:
```suggestion
            raise ValueError("HabanaDataLoader only supports pin_memory=True")
```

File: examples/pl_hpu/hpu_datamodule_sample.py [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+    model = RN50Module()
+
+    trainer.fit(model, datamodule=data_module)
+    trainer.validate(model, datamodule=data_module)
+    trainer.test(model, datamodule=data_module)
+
+
+if __name__ == "__main__":
+
+    parser = argparse.ArgumentParser(description="PyLight ImageNet Training")
```
Comment (kaushikb11) [inline examples/pl_hpu/hpu_datamodule_sample.py:122]:
```suggestion
    parser = argparse.ArgumentParser(description="Pytorch Lightning ImageNet Training")
```

File: src/pytorch_lightning/utilities/hpu_datamodule.py [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+
+    # Data loading code
+    dataset_train = torchvision.datasets.ImageFolder(traindir, train_transforms)
+
+    dataset_val = torchvision.datasets.ImageFolder(valdir, val_transforms)
+
+    return dataset_train, dataset_val
+
+
+class HPUDataModule(pl.LightningDataModule):
```
Comment (justusschock) [inline src/pytorch_lightning/utilities/hpu_datamodule.py:78]: IMO we shouldn't add this here. There are multiple reasons:

1.) This is too domain-specific. It only allows CV datasets (more specifically only imagenet so far). Lightning tries to be agnostic w.r.t. the task and data domain at hand.

2.) It is too accelerator specific. We want all the code to be hardware specific for the user and switching out the DM depending on the hardware isn't really agnostic. Usually hardware only reflects to accelerators and if necessary strategies and plugins. Personally, I'd prefer to keep this limitation.

3.) There are a ton of tools to accelerate loading for a specific usecase. There is nvidia dali, ffcv and many others. The interface of PL allows you to easily plugin each of those without requiring any additional code on our side. As far as I can tell, this would be the same case with the `habana_dataloader`. 

4.) IMO this does not really add a lot of benefits to the average user (mainly due to (1)) but only moves the habana imports to lightning imports.

**Review by justusschock — COMMENTED** [review]
Hey @jerome-habana 

Thanks for the implementation!

However, I am not sure whether I would really want to have this merged. Let's discuss!

cc @carmocca @awaelchli @Borda

File: src/pytorch_lightning/utilities/hpu_datamodule.py (hunk 11 above)
Comment (jerome-habana) [reply src/pytorch_lightning/utilities/hpu_datamodule.py:78]: Thanks @justusschock for the quick review. Points to consider
- there needs to be a place to maintain the extensible classes of datamodule per accelerator. Lightning should provide interfaces to support generic as well as accelerator specific parameters.
- I view datamodule as one more strategy/ plugins wrt accelerators. It need not be generic. If all strategies needs to be generic, we need to have a mechanism to ensure all plugins/ strategies work well with all available accelerators.
- AFAIK, dali doesn't support all datasets. We intend to enhance the supported data formats in the subsequent releases
- We need a mechanism in lightning to show case users on how to make use specific features. Error handling as part of it is only to make user life simple

Comment (carmocca) [reply src/pytorch_lightning/utilities/hpu_datamodule.py:78]: I agree with @justusschock. We don't want to add a `HPUDataModule` just as there isn't an `HPULightningModule` or `HPUTrainer`. If you want the implementation to live in this repository, we can choose one of the following two options instead:
- The strategy manages the DataLoader, as we do for IPUs (hacky solution but simple): https://github.com/Lightning-AI/lightning/blob/2b7cd58e06702e7fc96e65001a3853ed2aabc1bf/src/pytorch_lightning/strategies/ipu.py#L240-L254
- Implement a new mechanism to add plug-in support for custom DataLoaders, this would be the same for HPUs as for FFCV etc. For example, a mixin class. This will require a design discussion.

If you still prefer the simplicity of the HPUDataModule, this is totally fine but we would want that it is maintained inside the `hpu_dataloader` library. Where we update our docs to suggest it to HPU users.

Comment (jerome-habana) [reply src/pytorch_lightning/utilities/hpu_datamodule.py:78]: Thanks @carmocca . 
- I think a generic design is needed so that all accelerators can inherit & it create a new strategy for datamodule. I believe the proposed PR partially resolves it. 
- Poptorch method doesn't seem scalable.
- The changes cannot be maintained in hpu_dataloader as its specific to lightning plugins. 
- We already showcase pytorch users on how to make use of habana_Dataloader. I prefer a similar mechanism for lightning users. Just updating the document will not help a simple user I think.

File: examples/pl_hpu/hpu_datamodule_sample.py [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+    model = RN50Module()
+
+    trainer.fit(model, datamodule=data_module)
+    trainer.validate(model, datamodule=data_module)
+    trainer.test(model, datamodule=data_module)
+
+
+if __name__ == "__main__":
+
+    parser = argparse.ArgumentParser(description="Pytorch Lightning ImageNet Training")
```
Comment (Borda) [inline examples/pl_hpu/hpu_datamodule_sample.py:122]: Could we use LightningCLI?
cc: @carmocca
